File: skills/ads/scripts/ads_skill/ads/publish.py
"""The one public publish path: `publish_v1` (budget + campaign + N ad groups,
each with RSA + keywords; reuses an existing campaign of the same name). It
catches SDK errors at step granularity (via `_step`) and returns a RunOutcome
recording partial successes and the failing step. Revisions to live ads go
through ads.sh apply-fixes, not here.

Entity construction lives in entities.py; the step-error machinery in errors.py.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .entities import (
    _archive_campaigns_by_name,
    _create_ad_group,
    _create_callouts,
    _create_campaign_budget,
    _create_keywords,
    _create_negative_keywords,
    _create_price_asset,
    _create_responsive_search_ad,
    _create_search_campaign,
    _create_sitelinks,
    _create_structured_snippet,
    _find_existing_ad_group,
    _find_existing_campaign,
    _target_devices,
    _target_us_canada,
)
from .errors import _StepError, _sdk_version, _step
from ..lib.auth import load_client
from ..lib.schema import Brief, Failure


@dataclass
class ExecAdGroup:
    name: str
    adGroupId: str | None = None
    responsiveSearchAdId: str | None = None
    keywordResourceNames: tuple[str, ...] = ()


@dataclass
class ExecResults:
    """What publish_v1 created — returned to the caller for a run summary.
    Not persisted; the live account + Google change history are the record."""
    budgetId: str | None = None
    campaignId: str | None = None
    sitelinkResourceNames: tuple[str, ...] = ()
    calloutResourceNames: tuple[str, ...] = ()
    priceAssetResourceNames: tuple[str, ...] = ()
    structuredSnippetResourceNames: tuple[str, ...] = ()
    adGroups: list[ExecAdGroup] = field(default_factory=list)


@dataclass
class RunOutcome:
    results: ExecResults
    failure: Failure | None = None
    executor_version: str = "unknown"


# ---------- Public API ----------


def publish_v1(customer_id: str, brief: Brief, archive_existing: bool = False) -> RunOutcome:
    sdk_version = _sdk_version()
    results = ExecResults(adGroups=[ExecAdGroup(name=ag.name) for ag in brief.adGroups])
    try:
        client = load_client()
        if archive_existing:
            _step(
                "archive-existing-campaign",
                lambda: _archive_campaigns_by_name(client, customer_id, brief.campaign.name),
            )
        existing_campaign = None if archive_existing else _step(
            "find-existing-campaign",
            lambda: _find_existing_campaign(client, customer_id, brief),
        )
        if existing_campaign:
            results.campaignId, results.budgetId = existing_campaign
        else:
            results.budgetId = _step(
                "create-campaign-budget",
                lambda: _create_campaign_budget(client, customer_id, brief),
            )
            results.campaignId = _step(
                "create-search-campaign",
                lambda: _create_search_campaign(client, customer_id, brief, results.budgetId),
            )
            _step(
                "target-location",
                lambda: _target_us_canada(client, customer_id, results.campaignId),
            )
            _step(
                "target-devices",
                lambda: _target_devices(client, customer_id, results.campaignId, brief.campaign.devices),
            )
            _step(
                "create-negative-keywords",
                lambda: _create_negative_keywords(
                    client,
                    customer_id,
                    results.campaignId,
                    brief.campaign.negativeKeywords,
                ),
            )
            results.sitelinkResourceNames = tuple(
                _step(
                    "create-sitelinks",
                    lambda: _create_sitelinks(client, customer_id, brief, results.campaignId),
                )
            )
            results.calloutResourceNames = tuple(
                _step(
                    "create-callouts",
                    lambda: _create_callouts(client, customer_id, brief, results.campaignId),
                )
            )
            results.priceAssetResourceNames = tuple(
                _step(
                    "create-price-asset",
                    lambda: _create_price_asset(client, customer_id, brief, results.campaignId),
                )
            )
            results.structuredSnippetResourceNames = tuple(
                _step(
                    "create-structured-snippet",
                    lambda: _create_structured_snippet(client, customer_id, brief, results.campaignId),
                )
            )
        for idx, brief_ag in enumerate(brief.adGroups):
            slot = results.adGroups[idx]
            existing_ad_group = _step(
                "find-existing-ad-group",
                lambda b=brief_ag: _find_existing_ad_group(client, customer_id, b, results.campaignId),
                ad_group_name=brief_ag.name,
            )
            if existing_ad_group:
                slot.adGroupId = existing_ad_group
                should_create_keywords = False
            else:
                slot.adGroupId = _step(
                    "create-ad-group",
                    lambda b=brief_ag: _create_ad_group(client, customer_id, b, results.campaignId),
                    ad_group_name=brief_ag.name,
                )
                should_create_keywords = True
            slot.responsiveSearchAdId = _step(
                "create-responsive-search-ad",
                lambda b=brief_ag, rn=slot.adGroupId: _create_responsive_search_ad(
                    client, customer_id, b, rn
                ),
                ad_group_name=brief_ag.name,
            )
            if should_create_keywords:
                slot.keywordResourceNames = tuple(
                    _step(
                        "create-keywords",
                        lambda b=brief_ag, rn=slot.adGroupId: _create_keywords(
                            client, customer_id, b, rn
                        ),
                        ad_group_name=brief_ag.name,
                    )
                )
    except _StepError as exc:
        return RunOutcome(
            results,
            Failure(step=exc.step, message=exc.message, raw=exc.raw, adGroupName=exc.ad_group_name),
            sdk_version,
        )
    return RunOutcome(results, None, sdk_version)

File: skills/adkit/scripts/src/ads/publish.ts
/**
 * The one public publish path: {@link publishV1} (budget + campaign + N ad groups,
 * each with RSA + keywords; reuses an existing campaign of the same name). It
 * catches SDK errors at step granularity (via {@link step}) and returns a
 * {@link RunOutcome} recording partial successes and the failing step. Revisions to
 * live ads go through ads.sh apply-fixes, not here.
 *
 * Entity construction lives in entities.ts; the step-error machinery in errors.ts.
 *
 * Style note: everything in the ads layer is pure/functional, but {@link publishV1}
 * is the single deliberate exception — sequencing plus a mutable {@link ExecResults}
 * accumulator that records partial success is acceptable here (and only here),
 * because the port must, like the Python original, stop mid-sequence on the first
 * failing step while keeping whatever already succeeded. That can't be expressed as
 * a pure comprehension, so the mutation is isolated to this one orchestration edge.
 */

import {
  archiveCampaignsByName,
  createAdGroup,
  createCallouts,
  createCampaignBudget,
  createKeywords,
  createNegativeKeywords,
  createPriceAsset,
  createResponsiveSearchAd,
  createSearchCampaign,
  createSitelinks,
  createStructuredSnippet,
  findExistingAdGroup,
  findExistingCampaign,
  targetDevices,
  targetUsCanada,
} from "./entities.js";
import { StepError, sdkVersion, step } from "./errors.js";
import type { AdsClient } from "../lib/auth.js";
import type { Brief, Failure } from "../lib/schema.js";

/** Per-ad-group record of what {@link publishV1} created (or reused). */
export interface ExecAdGroup {
  name: string;
  adGroupId: string | null;
  responsiveSearchAdId: string | null;
  keywordResourceNames: readonly string[];
}

/**
 * What {@link publishV1} created — returned to the caller for a run summary. Not
 * persisted; the live account + Google change history are the record.
 */
export interface ExecResults {
  budgetId: string | null;
  campaignId: string | null;
  sitelinkResourceNames: readonly string[];
  calloutResourceNames: readonly string[];
  priceAssetResourceNames: readonly string[];
  structuredSnippetResourceNames: readonly string[];
  adGroups: ExecAdGroup[];
}

/** The full outcome of a publish run: what was created, plus the failure if any. */
export interface RunOutcome {
  results: ExecResults;
  failure: Failure | null;
  executorVersion: string;
}

/** Build a fresh {@link ExecAdGroup} slot for `name`, nothing created yet. */
export function makeExecAdGroup(name: string): ExecAdGroup {
  return { name, adGroupId: null, responsiveSearchAdId: null, keywordResourceNames: [] };
}

/** Build an empty {@link ExecResults}, one slot per ad group in `brief`. */
export function makeExecResults(brief: Brief): ExecResults {
  return {
    budgetId: null,
    campaignId: null,
    sitelinkResourceNames: [],
    calloutResourceNames: [],
    priceAssetResourceNames: [],
    structuredSnippetResourceNames: [],
    adGroups: brief.adGroups.map((ag) => makeExecAdGroup(ag.name)),
  };
}

/** Assemble a {@link RunOutcome} from its parts. */
export function makeRunOutcome(
  results: ExecResults,
  failure: Failure | null,
  executorVersion: string,
): RunOutcome {
  return { results, failure, executorVersion };
}

// ---------- Public API ----------

/**
 * Publish `brief` to `customerId` through `client`: create the campaign budget,
 * search campaign, its targeting + campaign-level assets, then each ad group with
 * its RSA and keywords. An existing campaign of the same name is reused (unless
 * `archiveExisting`, which archives same-named campaigns first and always creates
 * fresh). Newly-created ad groups get keywords; reused ones do not.
 *
 * The `client` is injected (the Python `publish_v1` called `load_client()`
 * internally) so this is unit-testable with a fake `AdsClient`; the bin/create
 * entrypoint calls {@link loadClient} and passes the result in.
 *
 * Never throws for an SDK/step failure: a {@link StepError} is caught and folded
 * into the returned {@link RunOutcome}'s `failure`, with `results` reflecting every
 * step that succeeded before it.
 */
export async function publishV1(
  client: AdsClient,
  customerId: string,
  brief: Brief,
  archiveExisting = false,
): Promise<RunOutcome> {
  const executorVersion = sdkVersion();
  const results = makeExecResults(brief);
  try {
    if (archiveExisting) {
      await step("archive-existing-campaign", () =>
        archiveCampaignsByName(client, customerId, brief.campaign.name),
      );
    }
    const existingCampaign = archiveExisting
      ? null
      : await step("find-existing-campaign", () => findExistingCampaign(client, customerId, brief));
    if (existingCampaign) {
      results.campaignId = existingCampaign[0];
      results.budgetId = existingCampaign[1];
    } else {
      results.budgetId = await step("create-campaign-budget", () =>
        createCampaignBudget(client, customerId, brief),
      );
      results.campaignId = await step("create-search-campaign", () =>
        createSearchCampaign(client, customerId, brief, results.budgetId!),
      );
      await step("target-location", () => targetUsCanada(client, customerId, results.campaignId!));
      await step("target-devices", () =>
        targetDevices(client, customerId, results.campaignId!, brief.campaign.devices),
      );
      await step("create-negative-keywords", () =>
        createNegativeKeywords(client, customerId, results.campaignId!, brief.campaign.negativeKeywords),
      );
      results.sitelinkResourceNames = await step("create-sitelinks", () =>
        createSitelinks(client, customerId, brief, results.campaignId!),
      );
      results.calloutResourceNames = await step("create-callouts", () =>
        createCallouts(client, customerId, brief, results.campaignId!),
      );
      results.priceAssetResourceNames = await step("create-price-asset", () =>
        createPriceAsset(client, customerId, brief, results.campaignId!),
      );
      results.structuredSnippetResourceNames = await step("create-structured-snippet", () =>
        createStructuredSnippet(client, customerId, brief, results.campaignId!),
      );
    }
    for (const [idx, briefAg] of brief.adGroups.entries()) {
      const slot = results.adGroups[idx]!;
      const existingAdGroup = await step(
        "find-existing-ad-group",
        () => findExistingAdGroup(client, customerId, briefAg, results.campaignId!),
        briefAg.name,
      );
      let shouldCreateKeywords: boolean;
      if (existingAdGroup) {
        slot.adGroupId = existingAdGroup;
        shouldCreateKeywords = false;
      } else {
        slot.adGroupId = await step(
          "create-ad-group",
          () => createAdGroup(client, customerId, briefAg, results.campaignId!),
          briefAg.name,
        );
        shouldCreateKeywords = true;
      }
      slot.responsiveSearchAdId = await step(
        "create-responsive-search-ad",
        () => createResponsiveSearchAd(client, customerId, briefAg, slot.adGroupId!),
        briefAg.name,
      );
      if (shouldCreateKeywords) {
        slot.keywordResourceNames = await step(
          "create-keywords",
          () => createKeywords(client, customerId, briefAg, slot.adGroupId!),
          briefAg.name,
        );
      }
    }
  } catch (exc) {
    if (exc instanceof StepError) {
      const failure: Failure = {
        step: exc.step,
        message: exc.message,
        raw: exc.raw,
        adGroupName: exc.adGroupName,
      };
      return makeRunOutcome(results, failure, executorVersion);
    }
    throw exc;
  }
  return makeRunOutcome(results, null, executorVersion);
}
